.NET SDK for Pluggable Components (Minimal State Store)

philliphoff opened the pull request:
This PR represents a first take on a .NET SDK for Dapr Pluggable Components, specifically, for state store pluggable components.  This includes only the components required for a "minimal" state store (which includes bulk APIs, but not queryable or transactional APIs).  Also excludes the test proxy component, as that doesn't really pertain to most of the discussion regarding the PR to date.

philliphoff committed:
Add infrastructure and API for state store Pluggable Components.

Signed-off-by: Phillip Hoff <phillip@orst.edu>

tmacam commented on src/Dapr.PluggableComponents.Protos/.gitignore:
```diff
@@ -0,0 +1,4 @@
+#
+# Protos are downloaded as part of build.
+#
+Protos/**/*.proto
```
TIL that you can drop .gitignore somewhere else besides the root of a project.

tmacam commented on src/Dapr.PluggableComponents.Protos/Dapr.PluggableComponents.Protos.csproj:
```diff
@@ -0,0 +1,46 @@
+    </PackageReference>
+  </ItemGroup>
+
+  <Target Name="DownloadComponentProtos" BeforeTargets="Build" Condition="!Exists('$(ProtosComponentsDir)')">
+    <MakeDir
+      Directories="$(ProtosComponentsDir)" />
+    <DownloadFile
+      SourceUrl="$(ProtosBaseUrl)/%(Protos.Identity)"
+      DestinationFolder="$(ProtosComponentsDir)" />
+  </Target>
```
Yeah!

Out of curiosity, how does this handle failure to download a file? Despite what its documentation says, the Java/Maven plugin for download does not abort on 404 and your project gets into a really strange state when download fails.

philliphoff replied:
While not explicitly documented, the `DownloadFile` task fails (and, thus, the build files) in the event of a `404`.  Still, I tweaked the download so that it will always attempt to download each file if it hasn't already, so the build will "recover" in the event of a temporary download failure for one of the files.

tmacam commented on src/Dapr.PluggableComponents.Protos/Dapr.PluggableComponents.Protos.csproj:
```diff
@@ -0,0 +1,46 @@
+  <PropertyGroup>
+    <ProtosVersion>v1</ProtosVersion>
+    <ProtosTag>v1.9.0</ProtosTag>
+    <ProtosBaseUrl>https://raw.githubusercontent.com/dapr/dapr/$(ProtosTag)/dapr/proto/components/$(ProtosVersion)</ProtosBaseUrl>
```
Don't you need `dapr/proto/common/v1/common.proto` too?

philliphoff replied:
No, the component `.proto` files do not refer to `common.proto`.  There *is* some overlap (e.g. the `Etag` definition is duplicated between the two) but I don't know if it make sense to try to consolidate; perhaps not now but as a separate wholistic gRPC API review.  That is, if we want to do that, it might make sense to have a larger refactoring to ensure that we have a hierarchy of "common" definitions that are the minimal necessary for each separate purpose (runtime vs. components) without pulling in definitions that are ultimately not used.

mcandeia replied:
please read: https://github.com/dapr/dapr/issues/5258

tmacam commented on src/Dapr.PluggableComponents.AspNetCore/DaprPluggableComponentsApplication.cs:
```diff
@@ -0,0 +1,179 @@
+        ((IDaprPluggableComponentsRegistrar)this).RegisterComponent(socketPath, context => context.ServiceProvider.GetRequiredService<TComponent>());
+    }
+
+    private readonly DaprPluggableComponentsRegistry registry = new DaprPluggableComponentsRegistry();
```
shouldn't this be together with the other attribute definitions above (lines 18-24).

philliphoff replied:
Fair enough. Updated.

tmacam commented on src/Dapr.PluggableComponents.AspNetCore/RegisteredComponentProvider.cs:
```diff
@@ -0,0 +1,48 @@
+    {
+        string socketPath = GetSocketPath(context);
+
+        var componentProvider = this.registry.GetComponentProvider<TComponent>(this.serviceProvider, socketPath);
+
+        if (componentProvider == null)
+        {
+            throw new InvalidOperationException($"Unable to obtain a provider for component type {typeof(TComponent)}.");
+        }
```
`?? throw`?

philliphoff replied:
Sure. Updated.

tmacam commented on src/Dapr.PluggableComponents.AspNetCore/RegisteredComponentProvider.cs:
```diff
@@ -0,0 +1,48 @@
+    {
+        var httpContext = context.GetHttpContext();
+        var socketFeature = httpContext.Features.Get<IConnectionSocketFeature>();
+        var socketPath = socketFeature?.Socket.LocalEndPoint?.ToString();
+
+        if (socketPath == null)
+        {
+            throw new InvalidOperationException("Unable to determine the socket on which a gRPC request was made.");
+        }
```
`?? throw`?

philliphoff replied:
Sure. Updated.

philliphoff commented:
@tmacam @mcandeia @johnewart

Switched from draft to PR.  Updates to feedback should all be addressed.  No significant architecture changes made since the last round of feedback, but the following addressed:

- Added XML docs for publicly-visible types
- Updated authorship/ownership/copyright metadata for NuGet packages and assemblies
- Added copyright headers to source files
- Made `.proto` download during build more robust

mcandeia commented on src/Dapr.PluggableComponents.Protos/Dapr.PluggableComponents.Protos.csproj:
```diff
@@ -0,0 +1,47 @@
+    </PackageReference>
+  </ItemGroup>
+
+  <Target Name="DownloadComponentProtos" BeforeTargets="Protobuf_BeforeCompile">
+    <MakeDir
+      Directories="$(ProtosComponentsDir)" />
+    <DownloadFile
+      Condition="!Exists('$(ProtosComponentsDir)/%(Protos.Identity)')"
+      SourceUrl="$(ProtosBaseUrl)/%(Protos.Identity)"
+      DestinationFolder="$(ProtosComponentsDir)" />
+  </Target>
```
amazing

philliphoff replied:
.NET's MSBuild infrastructure can sometimes be pretty powerful.

halspang commented on src/Dapr.PluggableComponents.AspNetCore/DaprPluggableComponentsApplication.cs:
```diff
@@ -0,0 +1,226 @@
+/// <summary>
+/// Represents an application that hosts Dapr Pluggable Components.
+/// </summary>
+public sealed class DaprPluggableComponentsApplication : IDaprPluggableComponentsRegistrar
```
Not saying I disagree with this, just curious. What benefit do we get from this over just providing an extension method for a standard web app like we do with Dapr?

philliphoff replied:
We had a lot of discussion around this earlier on. I had originally drafted the SDK as such (i.e. as an "extension" of an ASP.NET application), to be consistent with other ASP.NET SDKs and to provide a API familiar to an ASP.NET developer.  However, there was considerable concern amongst the group about customers grafting other non-pluggable-component-related functionality onto applications in ways we did not expect and, instead, a desire to keep applications built using the SDK very focused.

In the end we compromised a bit and shaped the SDK more like a Dapr pluggable component that *happens* to be implemented using ASP.NET, but exposing generic properties similar to that of an ASP.NET application (e.g. service collections, log managers, configuration) which would still be familiar to developers and in keeping with their current model of application development.

That said, I *also* keep many of the service configuration methods as public extension methods on the ASP.NET builder/application types in order to give developers the ability configure a "Dapr environment", should they need to customize beyond what the happy path provides, without having to completely start from scratch.

halspang commented on src/Dapr.PluggableComponents.AspNetCore/IDaprPluggableComponentsRegistrar.cs:
```diff
@@ -0,0 +1,27 @@
+
+namespace Dapr.PluggableComponents;
+
+internal interface IDaprPluggableComponentsRegistrar
```
Do we anticipate having a lot of these? If not, it may be good to move to abstract classes to allow for more flexibility.

I would also say that this is where we should have the docs and you can just use `<inheritdoc>` in the impls.

philliphoff replied:
No, I don't expect there to be more than one implementation as it's mostly intended to avoid a circular dependency (i.e. the application class and service builder class from knowing about each other) as well as to help restrict how the service builder interacts with the application class to avoid close coupling).

It doesn't have XML docs as it's an entirely internal type.  (I don't have a problem with documenting internal types, per se, but it wasn't convention in past teams/projects.)

halspang commented on src/Dapr.PluggableComponents.AspNetCore/WebApplicationBuilderExtensions.cs:
```diff
@@ -0,0 +1,107 @@
+    /// <param name="options">Options related to the creation of the socket file.</param>
+    /// <returns>The full path to the socketfile representing the service.</returns>
+    /// <exception cref="ArgumentException">Thrown if the socket file path cannot be built or has already been registered. </exception>
+    public static string AddDaprService(this WebApplicationBuilder builder, DaprPluggableComponentsServiceOptions options)
```
Should this just add the support services itself?

philliphoff replied:
This method is intended to be called multiple times, which would then cause those services to be added multiple times.  While I *fully expect* those service registration methods to reasonably handle such scenarios, as I don't own the code I can't *guarantee* that.  Hence, I'd say it's safer to keep them separate.

As these extension methods are intended for use by us, primarily, we can ensure the support services are added (and only once).  In cases where end users call this method, in setting up a customized Dapr pluggable component compatible environment beyond our happy path, I'd say the bar is a little higher in terms of understanding how these methods are expected to be used.

halspang commented on src/Dapr.PluggableComponents.AspNetCore/WebApplicationExtensions.cs:
```diff
@@ -0,0 +1,50 @@
+    /// <typeparam name="T">The type of gRPC service.</typeparam>
+    /// <param name="app">An ASP.NET endpoint builder.</param>
+    /// <returns>A <see cref="GrpcServiceEndpointConventionBuilder"/> through which the service can be further configured.</returns>
+    public static GrpcServiceEndpointConventionBuilder MapDaprPluggableComponentAdaptor<T>(this IEndpointRouteBuilder app)
```
Same question, should this register its support services?

philliphoff replied:
I would have the same argument as above regarding it not being certain, as we don't own that code, that it will handle multiple calls in the way we expect.  I'm *virtually certain* that it does, and we could even check, it being OSS, but that's still only a point in time certainty.

halspang commented on src/Dapr.PluggableComponents/Adaptors/StateStoreAdaptor.cs:
```diff
@@ -0,0 +1,247 @@
+        }
+        else
+        {
+            this.logger.LogInformation("Store does not support bulk operations; falling back to individual operations.");
```
Nit: This might be better as a debug as it could get pretty verbose if people are doing a lot of gets.

philliphoff replied:
Agree, changed all adaptor logging to "debug" level.

halspang commented on src/Dapr.PluggableComponents/Adaptors/StateStoreAdaptor.cs:
```diff
@@ -0,0 +1,247 @@
+        }
+        else
+        {
+            this.logger.LogInformation("Store does not support bulk operations; falling back to individual operations.");
```
Same here.

halspang commented on src/Dapr.PluggableComponents/Adaptors/StateStoreAdaptor.cs:
```diff
@@ -0,0 +1,247 @@
+        }
+        else
+        {
+            this.logger.LogInformation("Store does not support bulk operations; falling back to individual operations.");
```
Should probably be a debug since this still works fine and would be verbose.

halspang commented on src/Dapr.PluggableComponents/Components/IPluggableComponent.cs:
```diff
@@ -0,0 +1,29 @@
+/// <summary>
+/// Represents the root interface of all Dapr Pluggable Components.
+/// </summary>
+public interface IPluggableComponent
```
Would this plus the next 2 interfaces be better served as an abstract base class? We don't need to change it, but it seems like some default methods for ping/features would be cleaner than having each of these as an interface. If you rely on the interfaces for some type checking, that's also OK.

philliphoff replied:
The interfaces *are* used, in part, in the service registration process.  Still, I'd argue that an abstract base class, while a perfectly valid approach, doesn't provide significant benefit over interfaces here as there is no significant common implementation or logic that it could provide.  The developer must still override and provide complete implementations for each that will vary widely based on the type of component.  Given that, I feel (simple) interfaces provide more clarity in what is expected from the developer.

halspang reviewed (commented):
I had some questions about how/why we use a few things but overall this seems fine. My biggest concern is more with the usage of interfaces but it's really more of a personal preference and I recognize how they let us build some functionality that mirrors dapr runtime. 

So in the end, whatever the consensus is, I'm fine with. Just posting stuff for discussion's sake.

philliphoff commented:
@tmacam @mcandeia @halspang I think I've followed up on everyone's last comments.  Please let me know if there's anything else to discuss before a merge.  If not, as I don't have permissions, who can I ask to do so?

philliphoff commented:
Rebased/squashed commits to ensure signoff. No changes to content.

tmacam commented:
@artursouza, can you merge this one? I don't have permissions to do it :(
